api/index.php

```diff
@@ -0,0 +1,203 @@
+<?php
+$relPath="../pinc/";
+include_once($relPath.'bootstrap.inc');
+include_once($relPath.'User.inc');
+include_once($relPath.'misc.inc');
+include_once('exceptions.inc');
+include_once('v1.inc');
+
+// everything is a JSON response
+header("Content-Type: application/json");
+
+handle_cors_headers();
+
+if($maintenance)
+{
+    throw new ApiException("Site is in maintenance mode");
+}
+
+if(!$api_enabled)
+{
+    throw new ApiException("API is not enabled");
+}
+
+api();
+
+# ---------------------------------------------------------------------------
+
+function api()
+{
+    $username = api_authenticate();
+
+    api_rate_limit($username);
+
+    $query_params = $_GET;
+    $path = explode('/', array_get($query_params, "url", ""));
+    unset($query_params["url"]);
+
+    // if the path ended with a /, pop it off for correct routing
+    if($path[count($path) - 1] == '')
+    {
+        array_pop($path);
+    }
+
+    api_output_response(api_router($path, $query_params));
+}
+
+function api_authenticate()
+{
+    global $pguser;
+
+    $api_key = @$_SERVER['HTTP_X_API_KEY'];
+    if($api_key)
+    {
+        try {
+            $user = User::load_from_api_key($api_key);
+        } catch (NonexistentUserException $exception) {
+            throw new UnauthorizedError();
+        }
+        $pguser = $user->username;
+    }
+
+    if(!isset($pguser))
+    {
+        throw new UnauthorizedError();
+    }
+
+    return $pguser;
+}
+
+function api_rate_limit($key)
+{
+    global $api_rate_limit;
+    global $api_rate_limit_requests_per_window, $api_rate_limit_seconds_in_window;
+
+    if(!$api_rate_limit)
+    {
+        return;
+    }
+
+    $memcache = new Memcached();
+    $memcache->addServer('localhost', 11211);
+
+    // increment (or initialize) our count
+    $count = $memcache->get("$key:count");
+    $count = $count === FALSE ? 0 : $count + 1;
+    if($memcache->set("$key:count", $count) === FALSE)
```

Comment:
As a microoptimization, in the common case we could replace get/set with increment to avoid a second round trip...
https://www.php.net/manual/en/memcache.increment.php

Reply:
I started trying that, but increment only works when using the binary protocol (see first comment on that page) which it looks like the version of PHP we use on TEST doesn't support. I banged my head against this for longer than I want to admit.